File: src/main/kotlin/java/_02_data_calculations_actions/megaMart/functional/Main.java
package java._02_data_calculations_actions.megaMart.functional;

import java._02_data_calculations_actions.megaMart.givenCode.CartItem;
import java._02_data_calculations_actions.megaMart.givenCode.DOM;

import java.util.ArrayList;
import java.util.List;

import static java._02_data_calculations_actions.megaMart.givenCode.App.shoppingCart;
// After refactoring with calculation, the following import is no longer needed:
// import static _02_data_calculations_actions.megaMart.givenCode.App.shoppingCartTotal;


public class Main {

    public static List<CartItem> addItemToCart(List<CartItem> cart, String name, Double price) {
        // Instead of updating a global variable, we now pass in the cart as an input
        // and return a cart as an explicit output
        cart = addItem(cart, new CartItem(name, price));    // Updating/reading a global variable is an action
        calcCartTotal(cart);     // This is still an action, but it now uses an explicit input instead of reading a global                                              // The lines below fit better in this function than as a separate function

        return cart;
    }

    // This new function is a calculation, and thus easily testable and reusable
    // (We now use addItem in a couple places: addItemToCart, and in updateShippingIcons to test for free shipping)
    public static List<CartItem> addItem(List<CartItem> items, CartItem item) {
        var newItems = new ArrayList<>(items);
        newItems.add(item);
        return newItems;
    }

    public static void calcCartTotal(List<CartItem> cart) {
        var total = calcTotal(cart);
        DOM.setCartTotal(total);                    // Updating the DOM is an action
        updateShippingIcons(cart);          // This is still an action, but it now uses an explicit input
        updateTaxDom(total);
    }

    public static void updateShippingIcons(List<CartItem> cart) {
        var buyButtons = DOM.getBuyButtons();   // Reading the DOM is an action

        for (var button : buyButtons) {
            // There is a business rule embedded here (carts >= 20 get free shipping)
            // This code also requires the shoppingCartTotal to be initialized before testing
            if (getsFreeShipping(addItem(cart, button.item()))) {
                // Can't test this code without a mock DOM
                button.showFreeShippingIcon();      // Updating the DOM is an action
            } else {
                button.hideFreeShippingIcon();      // Updating the DOM is an action
            }
        }
    }

    public static void updateTaxDom(double shoppingCartTotal) {
        // By passing in the total we now have no external inputs
        // var tax = shoppingCartTotal * 0.13;  // Reading a global variable is an action
        DOM.setTax(calcTax(shoppingCartTotal));                     // Updating the DOM is still an action
    }

    // These are business rules encapsulated as calculations.
    // It is often useful to place business rule calculations like this into pure functions
    // so they may be easily reused.
    public static double calcTax(double amount) {
        return amount * 0.13;
    }

    public static boolean getsFreeShipping(List<CartItem> cart) {
        return calcTotal(cart) >= 20.0;
    }

    public static double calcTotal(List<CartItem> items) {
        var total = 0.0;
        for (CartItem item : items) {
            total += item.price();
        }
        return total;

        // Or, a few equivalent one-liners using streams:
//        return items.stream().map( item -> item.price() ).reduce(0.0, (a, b) -> a + b);
//        return items.stream().map(CartItem::price).reduce(0.0, Double::sum);
//        return items.stream().mapToDouble(CartItem::price).sum();
    }


}

File: src/main/kotlin/_02_data_calculations_actions/megaMart/functional/Main.kt
package _02_data_calculations_actions.megaMart.functional

data class CartItem(val name: String, val price: Double)


fun addItemToCart(cart: List<CartItem>, name: String, price: Double): List<CartItem> {
    // Instead of updating a global variable, we now pass in the cart as an explicit input
    // and return a new cart with the new item added
    val updatedCart = addItem(cart, CartItem(name, price))

    calcCartTotal(updatedCart)  // This is still an action, but it now uses an explicit input rather than reading a global

    return updatedCart

}

// This new function is a calculation, and thus easily testable and reusable
// (We now use addItem in a couple places: addItemToCart, and in updateShippingIcons to test for free shipping)
fun addItem(cart: List<CartItem>, item: CartItem) =
    cart + item  // Kotlin's + operator respects data immutability
                 // A new list is created with the new item added

fun calcCartTotal(cart: List<CartItem>) {
    val total = calcTotal(cart)
    setDOMCartTotal(total)       // Update the DOM is an action
    updateDOMShippingIcons(cart) // This too is still an action, but it now uses an explicit input rather than reading a global
    updateDOMTax(total)
}

fun updateDOMShippingIcons(cart: List<CartItem>) {
    val buyButtons = getDOMBuyButtons() // Accessing the DOM is an action
    for (button in buyButtons) {
        if (getsFreeShipping(addItem(cart, button.item))) {
            println("Free shipping icon shown on ${button.item.name}")
        } else {
            println("Free shipping icon hidden on ${button.item.name}")
        }
    }
}

fun setDOMCartTotal(total: Double) {
    // Update the total in the DOM (an action)
    // e.g. document.getElementById("cart-total").innerText = total
    // To keep things simple here, we'll just print the change
    println("Updated cart total to $total")
}

fun updateDOMTax(total: Double) {
    // val tax = shoppingCartTotal * 0.13  // Reading a global variable is an action
    // By passing in the total as an explicit input, we can now calculate the tax without relying on a global variable
    println("Updated tax to ${calcTax(total)}")
}

// These are business rules encapsulated as calculations.
// It is often useful to place business rule calculations like this into pure functions
// so they may be easily tested and reused.
fun calcTax(total: Double) = total * 0.13
fun getsFreeShipping(cart: List<CartItem>) = calcTotal(cart) > 20.0
fun calcTotal(cart: List<CartItem>) = cart.sumOf { it.price }
// The sumOf is a reduce function that iterates over the list and sums the result of the lambda function for each item
// The above one-liner is equivalent to:
// fun calcTotal(cart: List<CartItem>): Double {
//     var total = 0.0
//     for (item in cart) {
//         total += item.price
//     }
//     return total
// }




data class BuyButton(val item: CartItem)
fun getDOMBuyButtons() =
// NOTE: This is just a simulation of something like
// document.getElementsByClassName("buy-button")

    // To keep things simple here, we just simulate with a list of 3 buttons
    listOf(
        BuyButton(CartItem("Item 1", 10.0)),
        BuyButton(CartItem("Item 2", 15.0)),
        BuyButton(CartItem("Item 3", 5.0))
    )